File: java/sorting/BucketSortProduction.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public final class BucketSortProduction {

    private static final int DEFAULT_MAX_BUCKETS = 10_000;

    private BucketSortProduction() {
    }

    public static List<Double> bucketSort(List<Double> values, double bucketSize) {
        return bucketSort(values, bucketSize, DEFAULT_MAX_BUCKETS);
    }

    public static List<Double> bucketSort(List<Double> values, double bucketSize, int maxBuckets) {
        Objects.requireNonNull(values, "values");

        if (bucketSize <= 0.0 || !Double.isFinite(bucketSize)) {
            throw new IllegalArgumentException("bucketSize must be finite and greater than zero");
        }

        if (maxBuckets <= 0) {
            throw new IllegalArgumentException("maxBuckets must be greater than zero");
        }

        if (values.isEmpty()) {
            return new ArrayList<>();
        }

        for (double value : values) {
            if (!Double.isFinite(value)) {
                throw new IllegalArgumentException("bucketSort only accepts finite numeric values");
            }
        }

        double minimum = Collections.min(values);
        double maximum = Collections.max(values);
        int bucketCount = (int)((maximum - minimum) / bucketSize) + 1;

        if (bucketCount > maxBuckets) {
            List<Double> result = new ArrayList<>(values);
            Collections.sort(result);
            return result;
        }

        List<List<Double>> buckets = new ArrayList<>(bucketCount);

        for (int index = 0; index < bucketCount; index++) {
            buckets.add(new ArrayList<>());
        }

        for (double value : values) {
            int index = (int)((value - minimum) / bucketSize);
            buckets.get(Math.min(index, bucketCount - 1)).add(value);
        }

        List<Double> result = new ArrayList<>(values.size());

        for (List<Double> bucket : buckets) {
            Collections.sort(bucket);
            result.addAll(bucket);
        }

        return result;
    }
}

File: python/sorting/bucket_sort_production.py
from collections.abc import Iterable
from math import isfinite


def bucket_sort(
    values: Iterable[float],
    bucket_size: float = 5.0,
    *,
    max_buckets: int = 10_000,
) -> list[float]:
    items = list(values)

    if bucket_size <= 0 or not isfinite(bucket_size):
        raise ValueError("bucket_size must be finite and greater than zero")

    if max_buckets <= 0:
        raise ValueError("max_buckets must be greater than zero")

    if not items:
        return []

    if any(not isfinite(value) for value in items):
        raise ValueError("bucket_sort only accepts finite numeric values")

    minimum = min(items)
    maximum = max(items)
    bucket_count = int((maximum - minimum) / bucket_size) + 1

    if bucket_count > max_buckets:
        return sorted(items)

    buckets: list[list[float]] = [[] for _ in range(bucket_count)]

    for value in items:
        index = min(int((value - minimum) / bucket_size), bucket_count - 1)
        buckets[index].append(value)

    result: list[float] = []

    for bucket in buckets:
        bucket.sort()
        result.extend(bucket)

    return result
